### ISSUE TITLE
Can't instantiate final class extending ArrayIterator

### ISSUE THREAD

**Issue by ciaranmcnulty** [doctrine/instantiator#39]
```
final class Foo extends ArrayIterator
{
}

(new \Doctrine\Instantiator\Instantiator()->instantiate(Foo::class);
```

```
PHP Fatal error:  Uncaught Exception: unserialize(): Error at offset 13 of 14 bytes in /Users/cmcnulty/Documents/Code/phpspec/vendor/doctrine/instantiator/src/Doctrine/Instantiator/Exception/UnexpectedValueException.php:76
```

Reproduced on 1.0.5 and 1.1.0 and, not had a chance to debug. Was reported to PhpSpec as https://github.com/phpspec/phpspec/issues/1124

**Comment by Ocramius** [comment]
Is the class being final a required factor?

**Comment by ciaranmcnulty** [comment]
non-final doesn't reproduce the error

**Comment by ciaranmcnulty** [comment]
Yep final makes it go back to the unserialization strategy, which seems to have an issue with the ArrayIterator

**Comment by mikeSimonson** [comment]
@ciaranmcnulty This seems to be a hhvm serialazation bug. https://3v4l.org/1dYQD

**Comment by ciaranmcnulty** [comment]
Waaait, is it that it's working fine but the error handler is catching the notice?

**Comment by mikeSimonson** [comment]
note that with that class that extends ArrayIterator
This works in PHP and breaks in hhvm
´´´
unserialize('C:3:"Foo":0:{}')
´´´
And This works in hhvm and breaks in PHP
´´´
unserialize('O:3:"Foo":0:{}')
´´´

**Comment by Ocramius** [comment]
@mikeSimonson what is the resolution here?

**Comment by ciaranmcnulty** [comment]
I was wondering too :-)

If it’s “won’t fix” can you say explicitly so we can plan how we handle this error being thrown?

**Comment by Ocramius** [comment]
Needs clarification. I'm fine with dropping HHVM support and having a workaround just for `ArrayIterator` if anybody has an idea of how to get there :P

**Comment by ciaranmcnulty** [comment]
The fact it works differently in HHVM is a red herring IMO

the `final` is excluding the Reflection strategy for creating the object

The fact it extends ArrayObject means it somehow can't be deserialized from a string starting with `c:` but I don't know enough about serialization to suggest why (I tried to read http://www.phpinternalsbook.com/classes_objects/serialization.html#serializing-internal-objects and my eyes got tired)

**Comment by ciaranmcnulty** [comment]
As a side note I very much doubt specific serialization formats are at all going to be thought about in terms of BC when PHP changes things :-/

**Comment by mikeSimonson** [comment]
Sorry that it wasn't clear enough.

In my tests the only way to serialize the sample code in https://3v4l.org/1dYQD to 'O:3:"Foo":0:{}' is to use hhvm. And hhvm (HHVM-3.22.1) can unserialize that on my machine without any notice or error. https://3v4l.org/eX9TO

It just failed trying to unserialize hhvm serialization in php. Which seemed like a edge case to me.

That's the reason I closed it.

But maybe that's not was is going on in your case ?

**Comment by ciaranmcnulty** [comment]
I'm not using HHVM, no

**Comment by mikeSimonson** [comment]
Then, how did you ended up with the `'O:3:"Foo":0:{}'` serialized string ?

**Comment by ciaranmcnulty** [comment]
I was debugging and it was generated by Doctrine here:

https://github.com/doctrine/instantiator/blob/master/src/Doctrine/Instantiator/Instantiator.php#L104-L105

**Comment by gquemener** [comment]
This issue can now be closed :tada:

**Comment by Ocramius** [comment]
Fixed in #61
